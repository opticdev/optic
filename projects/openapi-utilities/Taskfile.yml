# https://taskfile.dev/#/installation
version: "3"

tasks:
  test:
    desc: Test
    cmds:
      - yarn run ws:test
