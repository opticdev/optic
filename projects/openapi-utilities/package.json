{
  "name": "@useoptic/openapi-utilities",
  "license": "MIT",
  "packageManager": "yarn@4.1.0",
  "version": "0.54.3",
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "files": [
    "/build",
    "/src"
  ],
  "scripts": {
    "build": "yarn tsc --build --verbose",
    "test": "ENVIRONMENT=test jest --colors",
    "clean": "rm -rf build"
  },
  "devDependencies": {
    "@babel/core": "^7.17.0",
    "@babel/plugin-transform-runtime": "^7.17.0",
    "@babel/preset-env": "^7.17.0",
    "@babel/preset-typescript": "^7.17.0",
    "@types/babel__core": "^7",
    "@types/babel__preset-env": "^7",
    "@types/is-url": "^1.2.30",
    "@types/js-yaml": "^4.0.5",
    "@types/json-stable-stringify": "^1.0.33",
    "@types/lodash.groupby": "^4.6.6",
    "@types/lodash.isequal": "^4.5.6",
    "@types/lodash.omit": "^4.5.7",
    "@types/node": "^18.0.0",
    "babel-jest": "^29.3.1",
    "jest": "^29.3.1",
    "prettier": "^3.0.0",
    "ts-jest": "^29.0.3",
    "ts-node": "^10.9.1",
    "typescript": "^5.0.0"
  },
  "dependencies": {
    "@useoptic/json-pointer-helpers": "workspace:*",
    "ajv": "^8.6.0",
    "ajv-errors": "~3.0.0",
    "ajv-formats": "~2.1.0",
    "chalk": "^4.1.2",
    "fast-deep-equal": "^3.1.3",
    "is-url": "^1.2.4",
    "js-yaml": "^4.1.0",
    "json-stable-stringify": "^1.0.1",
    "lodash.groupby": "^4.6.0",
    "lodash.isequal": "^4.5.0",
    "lodash.omit": "^4.5.0",
    "node-machine-id": "^1.1.12",
    "openapi-types": "^12.0.2",
    "ts-invariant": "^0.9.3",
    "url-join": "^4.0.1",
    "yaml-ast-parser": "^0.0.43"
  }
}
