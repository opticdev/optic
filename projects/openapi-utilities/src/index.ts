import {
  OpenAPITraverser,
  normalizeOpenApiPath,
} from './openapi3/implementations/openapi3/openapi-traverser';
import { OpenAPIV2, OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';
import { factsToChangelog } from './openapi3/sdk/facts-to-changelog';
export {
  FlatOpenAPIV2,
  FlatOpenAPIV3,
  FlatOpenAPIV3_1,
} from './flat-openapi-types';

import {
  ConceptualLocation,
  IChange,
  IFact,
  ILocation,
  OpenApiFact,
  OpenApiFieldFact,
  OpenApiBodyFact,
  OpenApiBodyExampleFact,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiParameterKind,
  OpenApiOperationFact,
  OpenApiRequestParameterFact,
  OpenApiResponseFact,
  OpenApiComponentSchemaExampleFact,
  OperationLocation,
  QueryParameterLocation,
  PathParameterLocation,
  HeaderParameterLocation,
  CookieParameterLocation,
  ResponseHeaderLocation,
  ResponseLocation,
  BodyLocation,
  BodyExampleLocation,
  ComponentSchemaLocation,
  FieldLocation,
  ChangeType,
  OpenApiSpecificationFact,
  OpenApi3SchemaFact,
  FactVariant,
  ChangeVariant,
  OpenApiKindToFact,
  OpenApiRequestFact,
} from './openapi3/sdk/types';
import {
  isFactVariant,
  isChangeVariant,
  isFactOrChangeVariant,
} from './openapi3/sdk/isType';
import {
  sourcemapReader,
  getSourcemapLink,
  GetSourcemapOptions,
} from './openapi3/implementations/openapi3/sourcemap-reader';

export { defaultEmptySpec } from './openapi3/constants';
export * from './openapi3/implementations/openapi3/types';
export * from './utilities/id';

export { SPEC_TAG_REGEXP, sanitizeGitTag } from './specs/tags';

export {
  getFactForJsonPath,
  constructFactTree,
  FactTree,
} from './openapi3/json-path-interpreters';

export { UserError } from './errors';
export { getEndpointsChanges } from './utilities/changelog';
export {
  BodyChange,
  ContentType,
  OpenApiEndpointChange,
  StatusCode,
  groupChangesAndRules,
} from './utilities/group-changes';
export { traverseSpec } from './utilities/traverse-spec';
export { compareChangesByPath } from './utilities/compare-changes-by-path';
export {
  getLabel,
  getOperationsChanged,
  getOperationsChangedLabel,
} from './utilities/count-changed-operations';
export * from './swagger2';
export {
  compareSpecs,
  CompareSpecResults,
} from './compare-specs/compare-specs';
export { groupDiffsByEndpoint } from './openapi3/group-diff';
export { ObjectDiff, typeofDiff, diff, reconcileDiff } from './diff/diff';
export {
  RuleResult,
  Severity,
  textToSev,
  sevToText,
  SeverityTextOptions,
  SeverityText,
} from './results';
export { OpenApiV3Traverser, OAS3 } from './openapi3/traverser';
export { isTruthyStringValue } from './utilities/truthy';
export {
  sourcemapReader,
  GetSourcemapOptions,
  getSourcemapLink,
  OpenApiFact,
  OpenAPITraverser,
  factsToChangelog,
  ConceptualLocation,
  IChange,
  OpenApiFieldFact,
  OpenApiBodyFact,
  OpenApiBodyExampleFact,
  OpenAPIV2,
  OpenAPIV3,
  OpenAPIV3_1,
  OpenApiKind,
  OpenApiParameterKind,
  OpenApiOperationFact,
  OpenApiHeaderFact,
  IFact,
  ILocation,
  OpenApiSpecificationFact,
  OpenApiRequestParameterFact,
  OpenApiResponseFact,
  OpenApiComponentSchemaExampleFact,
  OpenApi3SchemaFact,
  OperationLocation,
  QueryParameterLocation,
  PathParameterLocation,
  HeaderParameterLocation,
  CookieParameterLocation,
  ResponseHeaderLocation,
  ResponseLocation,
  BodyLocation,
  BodyExampleLocation,
  ComponentSchemaLocation,
  FieldLocation,
  ChangeType,
  FactVariant,
  ChangeVariant,
  OpenApiKindToFact,
  isFactVariant,
  isChangeVariant,
  isFactOrChangeVariant,
  OpenApiRequestFact,
};
export { normalizeOpenApiPath };
export * from './coverage/coverage';

export * from './types';

const packageJson = require('../package.json');
export const version = packageJson.version;
