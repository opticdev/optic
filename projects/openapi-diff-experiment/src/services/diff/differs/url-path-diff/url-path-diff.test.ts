import { pathParameterNamesForPathPattern, urlPathDiffFromSpec } from './index';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
import {
  OpenAPIDiffingQuestions,
  openApiDiffingQuestionsTestingStub,
} from '../../../read/types';

const gitHubOpenApiPaths = require('../../../../test/githubpaths.json');

const stub: OpenAPIDiffingQuestions = {
  ...openApiDiffingQuestionsTestingStub,
  paths(): string[] {
    return gitHubOpenApiPaths;
  },
};

const pathDiffer = urlPathDiffFromSpec(stub);

describe('url path differ', () => {
  it('will not match paths outside of set', async () => {
    const result = pathDiffer.compareToPath(
      OpenAPIV3.HttpMethods.GET,
      '/random-example'
    );
    expect(result).toMatchSnapshot();
  });

  it('can match /', async () => {
    const result = pathDiffer.compareToPath(OpenAPIV3.HttpMethods.GET, '/');
    expect(result).toMatchSnapshot();
  });

  it('will match concrete paths in set', async () => {
    const result = pathDiffer.compareToPath(
      OpenAPIV3.HttpMethods.GET,
      '/app/hook/config'
    );
    expect(result).toMatchSnapshot();
  });

  it('will match concrete paths with parameters set', async () => {
    const result = pathDiffer.compareToPath(
      OpenAPIV3.HttpMethods.GET,
      '/orgs/aidans-org/actions/runner-groups/group-5/runners/id-5'
    );
    expect(result).toMatchSnapshot();
  });
  it('will match concrete paths with parameters set 2', async () => {
    const result = pathDiffer.compareToPath(
      OpenAPIV3.HttpMethods.GET,
      '/repos/aidan-org/repo-1/pulls/15/reviews/23/dismissals'
    );
    expect(result).toMatchSnapshot();
  });

  it('will not match, but will find closest path with parameters set 2', async () => {
    const result = pathDiffer.compareToPath(
      OpenAPIV3.HttpMethods.GET,
      '/repos/aidan-org/repo-1/pulls/15/reviews/23/not-real'
    );
    expect(result).toMatchSnapshot();
  });

  /*
  known limitation -- nested ambiguous may not work properly
   */
  describe('ambiguous paths', () => {
    const stub: OpenAPIDiffingQuestions = {
      ...openApiDiffingQuestionsTestingStub,
      paths(): string[] {
        return ['/venues/top', '/venues/featured', '/venues/{venueId}'];
      },
    };

    const ambiguousPathDiffer = urlPathDiffFromSpec(stub);
    it('will match absolute path', () => {
      const result = ambiguousPathDiffer.compareToPath(
        OpenAPIV3.HttpMethods.GET,
        '/venues/top'
      );
      expect(result).toMatchSnapshot();
    });

    it('will match parameterized path', () => {
      const result = ambiguousPathDiffer.compareToPath(
        OpenAPIV3.HttpMethods.GET,
        '/venues/venue123'
      );
      expect(result).toMatchSnapshot();
    });
  });

  describe('path parameters from pattern', () => {
    it('can be extracted', () => {
      expect(
        pathParameterNamesForPathPattern('/example/{exampleId}/status/{action}')
      ).toMatchSnapshot();
    });
    it('empty if constants', () => {
      expect(
        pathParameterNamesForPathPattern('/example/list')
      ).toMatchSnapshot();
    });
  });
});
