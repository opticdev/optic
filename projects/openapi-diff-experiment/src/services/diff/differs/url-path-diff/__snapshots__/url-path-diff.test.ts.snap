// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`url path differ ambiguous paths will match absolute path 1`] = `
Object {
  "context": Object {
    "path": "/venues/top",
    "pathParameterValues": Object {},
    "urlPath": "/venues/top",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`url path differ ambiguous paths will match parameterized path 1`] = `
Object {
  "context": Object {
    "path": "/venues/{venueId}",
    "pathParameterValues": Object {
      "venueId": "venue123",
    },
    "urlPath": "/venues/venue123",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`url path differ can match / 1`] = `
Object {
  "context": Object {
    "path": "/",
    "pathParameterValues": Object {},
    "urlPath": "/",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`url path differ path parameters from pattern can be extracted 1`] = `
Array [
  "exampleId",
  "action",
]
`;

exports[`url path differ path parameters from pattern empty if constants 1`] = `Array []`;

exports[`url path differ will match concrete paths in set 1`] = `
Object {
  "context": Object {
    "path": "/app/hook/config",
    "pathParameterValues": Object {},
    "urlPath": "/app/hook/config",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`url path differ will match concrete paths with parameters set 1`] = `
Object {
  "context": Object {
    "path": "/orgs/{org}/actions/runner-groups/{runner_group_id}/runners/{runner_id}",
    "pathParameterValues": Object {
      "org": "aidans-org",
      "runner_group_id": "group-5",
      "runner_id": "id-5",
    },
    "urlPath": "/orgs/aidans-org/actions/runner-groups/group-5/runners/id-5",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`url path differ will match concrete paths with parameters set 2 1`] = `
Object {
  "context": Object {
    "path": "/repos/{owner}/{repo}/pulls/{pull_number}/reviews/{review_id}/dismissals",
    "pathParameterValues": Object {
      "owner": "aidan-org",
      "pull_number": "15",
      "repo": "repo-1",
      "review_id": "23",
    },
    "urlPath": "/repos/aidan-org/repo-1/pulls/15/reviews/23/dismissals",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`url path differ will not match paths outside of set 1`] = `
Object {
  "context": undefined,
  "diffs": Array [
    Object {
      "closestMatch": "/random-example",
      "method": "get",
      "path": "/random-example",
      "type": "UnmatchedPath",
    },
  ],
  "error": undefined,
  "isDiff": true,
  "isMatch": false,
}
`;

exports[`url path differ will not match, but will find closest path with parameters set 2 1`] = `
Object {
  "context": undefined,
  "diffs": Array [
    Object {
      "closestMatch": "/repos/{owner}/{repo}/pulls/{pull_number}/reviews/{review_id}/not-real",
      "method": "get",
      "path": "/repos/aidan-org/repo-1/pulls/15/reviews/23/not-real",
      "type": "UnmatchedPath",
    },
  ],
  "error": undefined,
  "isDiff": true,
  "isMatch": false,
}
`;
